caddy/shadowsocks_handler.go

```diff
@@ -114,11 +115,25 @@ func (h *ShadowsocksHandler) Provision(ctx caddy.Context) error {
 func (h *ShadowsocksHandler) Handle(cx *layer4.Connection, _ layer4.Handler) error {
 	switch conn := cx.Conn.(type) {
 	case transport.StreamConn:
-		h.service.HandleStream(cx.Context, conn)
+		h.service.HandleStream(cx.Context, &l4StreamConn{cx})
 	case net.PacketConn:
 		h.service.HandlePacket(conn)
 	default:
-		return fmt.Errorf("failed to handle unknown connection type: %t", conn)
+		return fmt.Errorf("failed to handle unknown connection type: %T", conn)
 	}
 	return nil
 }
+
+type l4StreamConn struct {
+	*layer4.Connection
+}
+
+var _ transport.StreamConn = (*l4StreamConn)(nil)
+
+func (c l4StreamConn) CloseWrite() error {
+	return nil
```

Comment:
This seems broken. We should have a way to send a FIN.
Does layer4.Connection not have a CloseWrite?

Reply:
It does not; it wraps a `net.Conn`. But actually, I think I should use the wrapped StreamConn's `CloseWrite()` and `CloseRead()` here.